python_modules/dagster/dagster/_core/definitions/decorators/decorator_assets_definition_builder.py

```diff
@@ -75,13 +75,39 @@ def get_function_params_without_context_or_config_or_resources(
     return new_input_args
 
 
-def build_named_ins(
+def validate_can_coexist(asset_in: AssetIn, asset_dep: AssetDep) -> None:
+    """Validates that the asset_in and asset_dep can coexist peacefully on the same asset key.
+    If both asset_in and asset_dep are set on the same asset key, expect that _no_ properties
+    are set on AssetIn except for the key itself.
+    """
+    if (
+        asset_in.metadata
+        or asset_in.key_prefix
+        or asset_in.dagster_type != NoValueSentinel
+        or asset_in.partition_mapping is not None
+    ):
+        raise DagsterInvalidDefinitionError(
+            f"Asset key '{asset_dep.asset_key.to_user_string()}' is used as both an input and a dependency. If an asset key is used as an input and also set as a dependency, the input should only define the relationship between the asset key and the input name, or optionally set the input_manager_key. Any other properties should either not be set, or should be set on the dependency."
```

Comment:
```suggestion
            f"Asset key '{asset_dep.asset_key.to_user_string()}' is used as both an input (AssetIn) and a dependency (AssetDep). If an asset key is used as an input and also set as a dependency, the input should only define the relationship between the asset key and the input name, or optionally set the input_manager_key. Any other properties should either not be set, or should be set on the dependency."
```

I think having an explicit mapping of `input` -> `AssetIn` and `dependency` -> `AssestDep` in the error message will be helpful to tell users exactly what class they can/cannot set these args on. Not set that this suggestion is the way to do that, just an option